### DISCUSSION TITLE
Error in installing PyBaMM

### DISCUSSION THREAD

**Discussion by Akila1993** [pybamm-team/PyBaMM#2731] [Q&A] [answered]
I tried installing PyBaMM in my Mac book Air. And getting following error message. 

Last login: Sat Feb 25 11:48:22 on ttys000
akilajayasinghe@Akilas-MacBook-Air ~ % $ /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install.sh)"
zsh: command not found: $
akilajayasinghe@Akilas-MacBook-Air ~ %  /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install.sh)" 
==> Checking for `sudo` access (which may request your password)...
Password:
==> This script will install:
/opt/homebrew/bin/brew
/opt/homebrew/share/doc/homebrew
/opt/homebrew/share/man/man1/brew.1
/opt/homebrew/share/zsh/site-functions/_brew
/opt/homebrew/etc/bash_completion.d/brew
/opt/homebrew

Press RETURN/ENTER to continue or any other key to abort:
==> /usr/bin/sudo /usr/sbin/chown -R akilajayasinghe:admin /opt/homebrew
==> Downloading and installing Homebrew...
HEAD is now at 0172ad102 Merge pull request #14794 from Bo98/api-check-shell
==> Downloading https://formulae.brew.sh/api/formula.json
######################################################################## 100.0%
==> Downloading https://formulae.brew.sh/api/cask.json
######################################################################## 100.0%
==> Installation successful!

==> Homebrew has enabled anonymous aggregate formulae and cask analytics.
Read the analytics documentation (and how to opt-out) here:
  https://docs.brew.sh/Analytics
No analytics data has been sent yet (nor will any be during this install run).

==> Homebrew is run entirely by unpaid volunteers. Please consider donating:
  https://github.com/Homebrew/brew#donations

==> Next steps:
- Run these two commands in your terminal to add Homebrew to your PATH:
    (echo; echo 'eval "$(/opt/homebrew/bin/brew shellenv)"') >> /Users/akilajayasinghe/.zprofile
    eval "$(/opt/homebrew/bin/brew shellenv)"
- Run brew help to get started
- Further documentation:
    https://docs.brew.sh

akilajayasinghe@Akilas-MacBook-Air ~ % (echo; echo 'eval "$(/opt/homebrew/bin/brew shellenv)"') >> /Users/akilajayasinghe/.zprofile
    eval "$(/opt/homebrew/bin/brew shellenv)"
akilajayasinghe@Akilas-MacBook-Air ~ % brew install python
Warning: python@3.11 3.11.2_1 is already installed and up-to-date.
To reinstall 3.11.2_1, run:
  brew reinstall python@3.11
akilajayasinghe@Akilas-MacBook-Air ~ % brew install python3
Warning: python@3.11 3.11.2_1 is already installed and up-to-date.
To reinstall 3.11.2_1, run:
  brew reinstall python@3.11
akilajayasinghe@Akilas-MacBook-Air ~ % brew install python3.9
Warning: No available formula with the name "python3.9". Did you mean python@3.9, python-tk@3.9, python@3.8, python@3.7, python@3.11, python@3.10 or python-yq?
==> Searching for similarly named formulae and casks...
==> Formulae
python@3.9 ✔        python@3.8          python@3.11 ✔       python-yq
python-tk@3.9       python@3.7          python@3.10

To install python@3.9 ✔, run:
  brew install python@3.9 ✔
akilajayasinghe@Akilas-MacBook-Air ~ % brew install python@3.9
Warning: python@3.9 3.9.16 is already installed and up-to-date.
To reinstall 3.9.16, run:
  brew reinstall python@3.9
akilajayasinghe@Akilas-MacBook-Air ~ % virtualenv env
zsh: command not found: virtualenv
akilajayasinghe@Akilas-MacBook-Air ~ % python3 -m venv env
akilajayasinghe@Akilas-MacBook-Air ~ % source env/bin/activate
(env) akilajayasinghe@Akilas-MacBook-Air ~ % brew install sundials
pip install pybamm
Warning: sundials 6.5.0_1 is already installed and up-to-date.
To reinstall 6.5.0_1, run:
  brew reinstall sundials
Collecting pybamm
  Using cached pybamm-0.2.4.post3-py3-none-any.whl (520 kB)
Requirement already satisfied: numpy>=1.16 in ./env/lib/python3.11/site-packages (from pybamm) (1.24.2)
Collecting scipy>=1.3
  Using cached scipy-1.10.1-cp311-cp311-macosx_12_0_arm64.whl (28.7 MB)
Requirement already satisfied: pandas>=0.24 in ./env/lib/python3.11/site-packages (from pybamm) (1.5.3)
Collecting anytree>=2.4.3
  Using cached anytree-2.8.0-py2.py3-none-any.whl (41 kB)
Collecting autograd>=1.2
  Using cached autograd-1.5-py3-none-any.whl (48 kB)
Collecting scikit-fem>=0.2.0
  Using cached scikit_fem-8.0.0-py3-none-any.whl (157 kB)
Collecting pybamm
  Using cached pybamm-0.2.4.post2-py3-none-any.whl (520 kB)
  Using cached pybamm-0.2.4.post1-py3-none-any.whl (520 kB)
  Using cached pybamm-0.2.4-py3-none-any.whl (520 kB)
  Using cached pybamm-0.2.3.post1.tar.gz (281 kB)
  Preparing metadata (setup.py) ... done
  Using cached pybamm-0.2.3-py3-none-any.whl (485 kB)
  Using cached pybamm-0.2.2.post3-py3-none-any.whl (450 kB)
  Using cached pybamm-0.2.2.post2-py3-none-any.whl (448 kB)
  Using cached pybamm-0.2.2.post1-py3-none-any.whl (446 kB)
  Using cached pybamm-0.2.2-py3-none-any.whl (444 kB)
  Using cached pybamm-0.2.1-py3-none-any.whl (402 kB)
  Using cached pybamm-0.2.0-py3-none-any.whl (371 kB)
  Using cached pybamm-0.1.0-py3-none-any.whl (310 kB)
ERROR: Cannot install pybamm==0.1.0, pybamm==0.2.0, pybamm==0.2.1, pybamm==0.2.2, pybamm==0.2.2.post1, pybamm==0.2.2.post2, pybamm==0.2.2.post3, pybamm==0.2.3, pybamm==0.2.3.post1, pybamm==0.2.4, pybamm==0.2.4.post1, pybamm==0.2.4.post2 and pybamm==0.2.4.post3 because these package versions have conflicting dependencies.

The conflict is caused by:
    pybamm 0.2.4.post3 depends on casadi>=3.5.0
    pybamm 0.2.4.post2 depends on casadi>=3.5.0
    pybamm 0.2.4.post1 depends on casadi>=3.5.0
    pybamm 0.2.4 depends on casadi>=3.5.0
    pybamm 0.2.3.post1 depends on casadi>=3.5.0
    pybamm 0.2.3 depends on casadi>=3.5.0
    pybamm 0.2.2.post3 depends on casadi>=3.5.0
    pybamm 0.2.2.post2 depends on casadi>=3.5.0
    pybamm 0.2.2.post1 depends on casadi>=3.5.0
    pybamm 0.2.2 depends on casadi>=3.5.0
    pybamm 0.2.1 depends on casadi>=3.5.0
    pybamm 0.2.0 depends on casadi>=3.5.0
    pybamm 0.1.0 depends on casadi>=3.5.0

To fix this you could try to:
1. loosen the range of package versions you've specified
2. remove package versions to allow pip attempt to solve the dependency conflict

ERROR: ResolutionImpossible: for help visit https://pip.pypa.io/en/latest/topics/dependency-resolution/#dealing-with-dependency-conflicts
(env) akilajayasinghe@Akilas-MacBook-Air ~ % pip install pybamm
Collecting pybamm
  Using cached pybamm-0.2.4.post3-py3-none-any.whl (520 kB)
Requirement already satisfied: numpy>=1.16 in ./env/lib/python3.11/site-packages (from pybamm) (1.24.2)
Collecting scipy>=1.3
  Using cached scipy-1.10.1-cp311-cp311-macosx_12_0_arm64.whl (28.7 MB)
Requirement already satisfied: pandas>=0.24 in ./env/lib/python3.11/site-packages (from pybamm) (1.5.3)
Collecting anytree>=2.4.3
  Using cached anytree-2.8.0-py2.py3-none-any.whl (41 kB)
Collecting autograd>=1.2
  Using cached autograd-1.5-py3-none-any.whl (48 kB)
Collecting scikit-fem>=0.2.0
  Using cached scikit_fem-8.0.0-py3-none-any.whl (157 kB)
Collecting pybamm
  Using cached pybamm-0.2.4.post2-py3-none-any.whl (520 kB)
  Using cached pybamm-0.2.4.post1-py3-none-any.whl (520 kB)
  Using cached pybamm-0.2.4-py3-none-any.whl (520 kB)
  Using cached pybamm-0.2.3.post1.tar.gz (281 kB)
  Preparing metadata (setup.py) ... done
  Using cached pybamm-0.2.3-py3-none-any.whl (485 kB)
  Using cached pybamm-0.2.2.post3-py3-none-any.whl (450 kB)
  Using cached pybamm-0.2.2.post2-py3-none-any.whl (448 kB)
  Using cached pybamm-0.2.2.post1-py3-none-any.whl (446 kB)
  Using cached pybamm-0.2.2-py3-none-any.whl (444 kB)
  Using cached pybamm-0.2.1-py3-none-any.whl (402 kB)
  Using cached pybamm-0.2.0-py3-none-any.whl (371 kB)
  Using cached pybamm-0.1.0-py3-none-any.whl (310 kB)
ERROR: Cannot install pybamm==0.1.0, pybamm==0.2.0, pybamm==0.2.1, pybamm==0.2.2, pybamm==0.2.2.post1, pybamm==0.2.2.post2, pybamm==0.2.2.post3, pybamm==0.2.3, pybamm==0.2.3.post1, pybamm==0.2.4, pybamm==0.2.4.post1, pybamm==0.2.4.post2 and pybamm==0.2.4.post3 because these package versions have conflicting dependencies.

The conflict is caused by:
    pybamm 0.2.4.post3 depends on casadi>=3.5.0
    pybamm 0.2.4.post2 depends on casadi>=3.5.0
    pybamm 0.2.4.post1 depends on casadi>=3.5.0
    pybamm 0.2.4 depends on casadi>=3.5.0
    pybamm 0.2.3.post1 depends on casadi>=3.5.0
    pybamm 0.2.3 depends on casadi>=3.5.0
    pybamm 0.2.2.post3 depends on casadi>=3.5.0
    pybamm 0.2.2.post2 depends on casadi>=3.5.0
    pybamm 0.2.2.post1 depends on casadi>=3.5.0
    pybamm 0.2.2 depends on casadi>=3.5.0
    pybamm 0.2.1 depends on casadi>=3.5.0
    pybamm 0.2.0 depends on casadi>=3.5.0
    pybamm 0.1.0 depends on casadi>=3.5.0

To fix this you could try to:
1. loosen the range of package versions you've specified
2. remove package versions to allow pip attempt to solve the dependency conflict

ERROR: ResolutionImpossible: for help visit https://pip.pypa.io/en/latest/topics/dependency-resolution/#dealing-with-dependency-conflicts

[Error message.pdf](https://github.com/pybamm-team/PyBaMM/files/10830459/Error.message.pdf)

Can someone assist me please.

**Comment by rtimms** [comment, marked as answer]
If you’re on M1 or M2 [this issue ](https://github.com/pybamm-team/PyBaMM/issues/2346) might help.
